### DISCUSSION TITLE
client_bounding_boxes.py : remove non visible vehicles

### DISCUSSION THREAD

**Discussion by MMNavetty** [carla-simulator/carla#5552] [Q&A]
Hello Everyone,
I need to use vehicles bounding boxes like in example "client_bounding_boxes.py", but i need to delete all vehicles that are not visible by the sensor.
Is there a way to raycast in Carla ? Maybe i can use an obstacle detector for this application but i'm not sure if this will really work.
Thanks in advance,
MM

**Comment by MMNavetty** [comment]
ok, following [#314](https://github.com/carla-simulator/carla/issues/314#issuecomment-379788542), i think i'll try this method.

**Comment by yjzhai-cs** [comment]
> Hello Everyone, I need to use vehicles bounding boxes like in example "client_bounding_boxes.py", but i need to delete all vehicles that are not visible by the sensor. Is there a way to raycast in Carla ? Maybe i can use an obstacle detector for this application but i'm not sure if this will really work. Thanks in advance, MM

Hello, I have a similar problem. How do you remove the bounding box of a vehicle that is not visible to the sensor?

**Reply by MMNavetty** [reply]
I use segmentation camera, looking at pixels of the bbox if it is the colour of the bbox's class then i keep the ground truth.
Depending on your version and classes detected you might need to change the pixels values checked. And you need to add a segmentation camera for every cam.

`


def is_box_visible(box, sensor_manager):

    # check 5 pixel to be sure !
    seg_pixel_1 = sensor_manager.last_seg_img[int((box[2]+box[3])/2), int((box[0]+box[1])/2)]
    seg_pixel_2 = sensor_manager.last_seg_img[int(box[2]+(box[3]-box[2])*2/3), int(box[0]+(box[1]-box[0])*2/3)]
    seg_pixel_3 = sensor_manager.last_seg_img[int(box[2]+(box[3]-box[2])/3), int(box[0]+(box[1]-box[0])/3)]
    seg_pixel_4 = sensor_manager.last_seg_img[int(box[2]+(box[3]-box[2])*2/3), int(box[0]+(box[1]-box[0])/3)]
    seg_pixel_5 = sensor_manager.last_seg_img[int(box[2]+(box[3]-box[2])/3), int(box[0]+(box[1]-box[0])*2/3)]

    visible1 = (seg_pixel_1[0]==60 and seg_pixel_1[1]==20 and seg_pixel_1[2]==220) or (seg_pixel_1[0]==142 and seg_pixel_1[1]==0 and seg_pixel_1[2]==0)
    visible2 = (seg_pixel_2[0]==60 and seg_pixel_2[1]==20 and seg_pixel_2[2]==220) or (seg_pixel_2[0]==142 and seg_pixel_2[1]==0 and seg_pixel_2[2]==0)
    visible3 = (seg_pixel_3[0]==60 and seg_pixel_3[1]==20 and seg_pixel_3[2]==220) or (seg_pixel_3[0]==142 and seg_pixel_3[1]==0 and seg_pixel_3[2]==0)
    visible4 = (seg_pixel_4[0]==60 and seg_pixel_4[1]==20 and seg_pixel_4[2]==220) or (seg_pixel_4[0]==142 and seg_pixel_4[1]==0 and seg_pixel_4[2]==0)
    visible5 = (seg_pixel_5[0]==60 and seg_pixel_5[1]==20 and seg_pixel_5[2]==220) or (seg_pixel_5[0]==142 and seg_pixel_5[1]==0 and seg_pixel_5[2]==0)

    return (visible1 or visible2 or visible3 or visible4 or visible5)
`
